qrsc/qrsc.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH QRSC "1" "January 2010" "qrsc (QRemoteSignal interface compiler) 0.7.0svn" "User Commands"
.SH NAME
qrsc \- manual page for qrsc (QRemoteSignal interface compiler) 0.7.0svn
.SH SYNOPSIS
.B qrsc
[\fIoptions\fR] \fIinterface.xml\fR
.SH DESCRIPTION
QRemoteSignal interface compiler. Creates client or service source files from interface XML description. One of \fB\-\-service\fR or \fB\-\-client\fR flags should always be specified but not both.
.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-client\fR
Create client class
.TP
\fB\-\-header\fR header
Specify output header file
.TP
\fB\-h\fR, \fB\-\-help\fR
Print this help and exit
.TP
\fB\-\-qt\-version\fR
Print Qt version information and exit
.TP
\fB\-s\fR, \fB\-\-service\fR
Create service class
.TP
\fB\-\-source\fR source
Specify output C++ source file
.TP
\fB\-u\fR, \fB\-\-update\fR update
Update interface file to new format version and save it in the file specified by this option parameter
.TP
\fB\-v\fR, \fB\-\-version\fR
Print version information and exit


ChangeLog
SVN Version
	* Added singleton factory for serializers. Explicit serializer
	creation is no more required.
	* Serializer test suit skips test instead of fail it if serializer
	doesn't suport argument type. (The only type that any serializer must
	support are QString and int)
	* DeviceManager class removed from library public interface. Use
	ServicesManager::addDevice instead of it.
	* Added signal clientError which is emited if received message is
	incorrect.
	* Added serializer versioning support.
	* Added QDataStream based serializer.

2009-11-30 Version 0.7.0
	* Build issues on windows are fixed.
	* Added ability to work with QIODevice to the ServicesManager class.
	* Added man page for qrsc utility. This manual page is generated by
	help2man program.
	* Improved error messages handling.
	* <method> element in the interface description is changed to <slot>.
	Automtic schema update can be done with qrsc utility.
	* Added <signal> elemnt to support remote signal/slot calls in both
	directions.
	* Added doxygen comments covering all library public API.
	* Written tutorial of library ussage.
	* Added qmake configuration feature file to build projects using
	QRemoteSignal.

2009-10-14 Version 0.6.0
	* Added support for the folowing types: QChar, char, signed char,
	unsigned char, short, unsigned short, unsigned, long, unsigned long,
	long long, unsigned long long, bool, QList<T>, QMap<QString,T> (where
	T is any type supported by library).
	* Added class to read write messages using any QIODevice (for example
	QTcpSocket, QUdpSocket...).
	* Added scons arguments caching.
	* Created test suit for testing message serializers.
	* Added INSTALL file with library build and installation instructions.

2009-09-13 Version 0.5.0
	* Added support of JSON as message protocol.
	* Added support of QString and int arguments.
	* Library core created.


qrsc/qrsc.cpp
/**
 * @file xic.cpp
 * @brief Entry point for XML interface compiler
 *
 * @author VestniK (Sergey N.Vidyuk) sir.vestnik@gmail.com
 * @date 29 Jul 2009
 */
#include <cstdio>

#include <QtCore/QTextStream>
#include <QtCore/QString>
#include <QtCore/QCoreApplication>

#include "interfacedocument.h"
#include "interfacecompiler.h"
#include "argvparser.h"
#include "config.h"

const QString HELP_FLAG = "help";
const QString VERSION_FLAG = "version";
const QString QT_VERSION_FLAG = "qt-version";
const QString SERVICE_FLAG = "service";
const QString CLIENT_FLAG = "client";
const QString HEADER_OPTION = "header";
const QString SOURCE_OPTION = "source";
const QString UPDATE_OPTION = "update";

int main(int argc, char *argv[]) {
   // Application Info
   QCoreApplication app(argc,argv);
   ArgvParser conf;
   app.setApplicationName(conf.tr("QRemoteSignal interface compiler"));
   app.setApplicationVersion(VERSION);
   QTextStream out(stdout,QIODevice::WriteOnly);
   QTextStream err(stderr,QIODevice::WriteOnly);
   // Command line options configuration
   conf.setAppDescription(conf.tr("QRemoteSignal interface compiler. Creates client or service source files from interface XML description. One of --%1 or --%2 flags should always be specified but not both.").arg(SERVICE_FLAG).arg(CLIENT_FLAG));
   conf.addFlag(HELP_FLAG,conf.tr("Print this help and exit"),'h');
   conf.addFlag(VERSION_FLAG,conf.tr("Print version information and exit"),'v');
   conf.addFlag(QT_VERSION_FLAG,conf.tr("Print Qt version information and exit"));

   conf.addFlag(SERVICE_FLAG,conf.tr("Create service class"),'s');
   conf.addFlag(CLIENT_FLAG,conf.tr("Create client class"),'c');
   conf.addOption(UPDATE_OPTION,conf.tr("Update interface file to new format version and save it in the file specified by this option parameter"),'u');
   conf.addOption(HEADER_OPTION,conf.tr("Specify output header file"));
   conf.addOption(SOURCE_OPTION,conf.tr("Specify output C++ source file"));
   if ( ! conf.parse() ) {
      err << conf.errorMessage() << endl;
      out << conf.helpStr();
      return 1;
   }
   // Print info if requested
   if ( conf.flags()[HELP_FLAG] ) {
      out << conf.helpStr();
      return 0;
   }
   if ( conf.flags()[QT_VERSION_FLAG] ) {
      out << conf.qtVersionStr();
      return 0;
   }
   if ( conf.flags()[VERSION_FLAG] ) {
      out << conf.versionStr();
      return 0;
   }
   if ( conf.arguments().isEmpty() ) {
      err << conf.tr("Error: Input file not specified!") << endl;
      out << conf.helpStr();
      return 1;
   }

   InterfaceDocument inputDoc( conf.arguments().first() );
   // Update document
   if ( ! conf.options()[UPDATE_OPTION].isEmpty() ) {
      QFile out(conf.options()[UPDATE_OPTION]);
      out.open(QIODevice::WriteOnly | QIODevice::Text);
      inputDoc.getIODevice()->seek(0);
      out.write( inputDoc.getIODevice()->readAll() );
      return 0;
   }
   if ( ! inputDoc.isValid() ) {
      err << inputDoc.error() << endl;
      return 1;
   }
   // Compile document
   InterfaceCompiler compiler(&inputDoc);

   /// @todo move output files to the compiler class.
   QString header( conf.options()[HEADER_OPTION] );
   QString source( conf.options()[SOURCE_OPTION] );
   if ( conf.flags()[SERVICE_FLAG] && conf.flags()[CLIENT_FLAG] ) {
      err << conf.tr("You should specify --%1 or --%2 flag but not both").arg(SERVICE_FLAG).arg(CLIENT_FLAG) << endl;
      return 1;
   }
   if ( conf.flags()[SERVICE_FLAG] ) {
      if ( ! header.isEmpty() ) {
         inputDoc.setServiceHeader(header);
      }
      if ( ! source.isEmpty() ) {
         inputDoc.setServiceSource(source);
      }
      if ( !compiler.compileServiceHeader() ) {
         err << app.tr("Failed to compile interface") << endl;
         return 1;
      }
      if ( !compiler.compileServiceSource() ) {
         err << app.tr("Failed to compile interface") << endl;
         return 1;
      }
   } else if ( conf.flags()[CLIENT_FLAG] ) {
      if ( ! header.isEmpty() ) {
         inputDoc.setClientHeader(header);
      }
      if ( ! source.isEmpty() ) {
         inputDoc.setClientSource(source);
      }
      if ( !compiler.compileClientHeader() ) {
         err << app.tr("Failed to compile interface") << endl;
         return 1;
      }
      if ( !compiler.compileClientSource() ) {
         err << app.tr("Failed to compile interface") << endl;
         return 1;
      }
   } else {
      err << app.tr("You should specify --%1 or --%2 flag").arg(SERVICE_FLAG).arg(CLIENT_FLAG) << endl;
      return 1;
   }

   return 0;
}
